source/client-side-encryption/tests/README.rst

```diff
@@ -0,0 +1,296 @@
+
+Prose Tests
+===========
+
+Tests for the KeyVault type are not included as part of the YAML tests. Tests are described
+as follows.
+
+#. Test creating a data key with the "local" KMS provider.
+
+   - Create a `KeyVault` with a "local" KMS provider.
+   - Create a data key with the "local" KMS provider using `KeyVault.createDataKey()`.
+   - Expect a BSON binary with subtype 4 to be returned.
+   - Expect a `findOne` on the key vault collection with `_id` set to the returned binary to return a document.
+   - Expect that document to have "masterKey.provider" set to "local"
+
+#. Test creating a data key with the "aws" KMS provider.
+
+   - Create a `KeyVault` with a "aws" KMS provider.
+   - Create a data key with the "aws" KMS provider using `KeyVault.createDataKey()`.
+   - Expect a BSON binary with subtype 4 to be returned.
+   - Expect a `findOne` on the key vault collection with `_id` set to the returned binary to return a document.
+   - Expect that document to have "masterKey.provider" set to "aws"
+
+#. Test explicit encrypt and decrypt with the "local" KMS provider.
+
+   - Create a `KeyVault` with a "local" KMS provider.
+   - Insert the a key document in to the key vault.
+   - Use `KeyVault.encrypt` to encrypt the value "hello" with the following:
+
+     - the algorithm "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
+     - a fixed 16 byte initialization vector
+     - the "local" KMS provider
+
+   - Expect the value is equal to a known BSON binary of subtype 6
+   - Use `KeyVault.decrypt` to decrypt the encrypted value
+   - Expect the value is equal to the string "hello"
+
+#. Test explicit encrypt and decrypt with the "aws" KMS provider.
+
+   - Create a `KeyVault` with a "aws" KMS provider.
+   - Insert the a key document in to the key vault.
+   - Use `KeyVault.encrypt` to encrypt the value "hello" with the following:
+
+     - the algorithm "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic"
+     - a fixed 16 byte initialization vector
+     - the "aws" KMS provider
+
+   - Expect the value is equal to a known BSON binary of subtype 6
+   - Use `KeyVault.decrypt` to decrypt the encrypted value
+   - Expect the value is equal to the string "hello"
+
+#. Test explicit encrypt of invalid values.
+
+   - Create a `KeyVault` with either a "local" or "aws" KMS provider
+   - Use `KeyVault.encrypt` to attempt to encrypt each BSON type with deterministic encryption.
+
+     - Expect a `string` to succeed. An exception MUST be thrown.
+     - Expect a `document` to fail. An exception MUST be thrown.
+     - Expect a `null` to fail. An exception MUST be thrown.
+     - Expect a BSON binary subtype 6 to fail. An exception MUST be thrown.
+
+   - Use `KeyVault.encrypt` to attempt to encrypt a document using randomized encryption.
+
+     - Expect a `document` to succeed.
+     - Expect a BSON binary subtype 6 to fail. An exception MUST be thrown.
+     - Expect a `null` to fail. An exception MUST be thrown.
```

Comment:
This concerns me.  Since null is treated almost the same as the absence of the key, I could see situations where applications inadvertently include a null value for an encrypted field.  Perhaps drivers should just silently not encrypt null.  Consider a case like this:

    MongoDB Enterprise > db.test.insert({_id : 1})
    WriteResult({ "nInserted" : 1 })
    MongoDB Enterprise > db.test.find({y : null})
    { "_id" : 1 }

This is a natural way that applications search for absence of a field, but if `y` is an encrypted field, this query will fail because the driver will throw trying to encrypt the query filter.

Feel free to resolve this if you'd prefer to deal with it at the spec level

Reply:
Oh, interesting. I'm not sure what the best decision is, but this is probably either a spec question, and perhaps one the query team should weigh in on. This would require a change in mongocryptd to ignore `null` on encrypted fields instead of erroring.

Regardless of that decision, perhaps explicit encryption should still error, because if the user is attempting to explicitly encrypt `null`, they could just pass `null` directly. What do you think?

Reply:
Reflecting more about this, perhaps it's not worth worrying about.  Applications can use `$exists` in filters, and $ifNull in aggregation expressions, and those will do the right thing without requiring the use of null